File: src/main/java/org/magicghostvu/actorvt/ActorRef.java
package org.magicghostvu.actorvt;

import org.jspecify.annotations.Nullable;
import org.magicghostvu.actorvt.context.GeneralActorContext;

public final class ActorRef<T> {

    public final String name;
    public final String path;

    @Nullable
    public volatile GeneralActorContext<T> context;


    public ActorRef(String name, GeneralActorContext<T> context) {
        this.name = name;
        this.context = context;
        this.path = context.getPath();
    }

    public void tell(T message) {
        GeneralActorContext<T> c = context;
        if (c != null && c.active) {
            try {
                c.messageQueue.put(message);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException("interrupted while sending to actor " + name, e);
            }
        } else {
            throw new IllegalArgumentException(String.format(
                    "actor %s with path %s not ready to receive msg (active: %s)",
                    name, path, c != null ? c.active : "null"));
        }
    }

    public boolean trySend(T message) {
        GeneralActorContext<T> c = context;
        if (c != null && c.active) {
            try {
                return c.messageQueue.offer(message);
            } catch (IllegalStateException ie) {
                return false;
            }
        } else throw new IllegalArgumentException(String.format(
                "actor %s with path %s not ready to receive msg (active: %s)",
                name, path, c != null ? c.active : "null"));

    }
}

File: src/main/kotlin/org/magicghostvu/actorvt/ActorRef.kt
package org.magicghostvu.actorvt


import org.magicghostvu.actorvt.context.GeneralActorContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory

// nên cho context thành nullable để detach khỏi actor ref ngay sau khi actor bị stop??
// ở cài đặt hiện tại nếu user có một actor đã bị xoá nhưng mà vẫn giữ ref thì data của actor đó vẫn còn trong memory
class ActorRef<in Protocol> internal constructor(
    val name: String,
    internal var context: GeneralActorContext<@UnsafeVariance Protocol>?,
) {

    lateinit var path: String

    init {
        val c = context
        if (c != null) {
            path = c.path
        }
    }

    // end user will use this
    fun tell(message: Protocol) {
        val c = context
        if (c != null && c.active) {
            c.messageQueue.put(message as Any)
        } else {
            logger.warn("actor {} not active so can not send msg {}", path, message)
        }
    }

    private companion object {
        private val logger: Logger = LoggerFactory.getLogger("actor-vt")
    }
}